sklearn/preprocessing/_encoders.py

```diff
@@ -97,7 +97,12 @@ def _fit(
                 else:
                     cats = result
             else:
-                cats = np.array(self.categories[i], dtype=Xi.dtype)
+                if np.issubdtype(Xi.dtype, np.str_):
+                    Xi_dtype = object
+                else:
+                    Xi_dtype = Xi.dtype
```

Comment:
What wold be nice is to have a way to find a "common dtype" where it could hold both data from categories and X I think.

Reply:
I tried `np.result_type()` which would do that. Didn't go with it in the end because there seems to be this special casing of "string like numpy? then object!" in `_check_X`. Any idea how we could find out what the rationale behind that is?

Reply:
yeah the code is moved around quite a bit, can't tell what the origin is. cc'ing @thomasjpfan here.

I think `result_type` should be a good enough option though.

Reply:
For string values of categorical variables, I think it's better to always use the object dtype (the only available representation for variable length strings stored in numpy arrays) .

Reply:
hopefully my worry about performance regression here is unfounded 😝 

I feel like we're adding a bunch of pointer lookups here. There's a big difference between object dtypes and other dtypes in numpy. Object dtypes can't even be persisted with numpy format.

Reply:
The arrays in the `.categories_` are small. And most of the time we access such an attribute we want to manipulate `str` objects in the end (e.g. to display category labels for a pipeline inspection tool.